docs/sources/flow/reference/components/prometheus.scrape.md

```diff
@@ -44,30 +44,31 @@ The following arguments are supported:
+`job_name`                    | `string`   | The value to use for the job label if not already set. | component name | no
+`extra_metrics`               | `bool`     | Whether extra metrics should be generated for scrape targets. | `false` | no
 `enable_protobuf_negotiation` | `bool`     | Whether to enable protobuf negotiation with the client. | `false` | no
-`honor_labels`             | `bool`     | Indicator whether the scraped metrics should remain unmodified. | `false` | no
-`honor_timestamps`         | `bool`     | Indicator whether the scraped timestamps should be respected. | `true` | no
-`params`                   | `map(list(string))` | A set of query parameters with which the target is scraped. | | no
-`scrape_classic_histograms` | `bool`     | Whether to scrape a classic histogram that is also exposed as a native histogram. | `false` | no
-`scrape_interval`          | `duration` | How frequently to scrape the targets of this scrape configuration. | `"60s"` | no
-`scrape_timeout`           | `duration` | The timeout for scraping targets of this configuration. | `"10s"` | no
-`metrics_path`             | `string`   | The HTTP resource path on which to fetch metrics from targets. | `/metrics` | no
-`scheme`                   | `string`   | The URL scheme with which to fetch metrics from targets. | | no
-`body_size_limit`          | `int`      | An uncompressed response body larger than this many bytes causes the scrape to fail. 0 means no limit. | | no
-`sample_limit`             | `uint`     | More than this many samples post metric-relabeling causes the scrape to fail | | no
-`target_limit`             | `uint`     | More than this many targets after the target relabeling causes the scrapes to fail. | | no
-`label_limit`              | `uint`     | More than this many labels post metric-relabeling causes the scrape to fail. | | no
-`label_name_length_limit`  | `uint`     | More than this label name length post metric-relabeling causes the scrape to fail. | | no
-`label_value_length_limit` | `uint`     | More than this label value length post metric-relabeling causes the scrape to fail. | | no
-`bearer_token` | `secret` | Bearer token to authenticate with. | | no
-`bearer_token_file` | `string` | File containing a bearer token to authenticate with. | | no
-`proxy_url` | `string` | HTTP proxy to proxy requests through. | | no
-`follow_redirects` | `bool` | Whether redirects returned by the server should be followed. | `true` | no
-`enable_http2` | `bool` | Whether HTTP2 is supported for requests. | `true` | no
+`honor_labels`                | `bool`     | Indicator whether the scraped metrics should remain unmodified. | `false` | no
+`honor_timestamps`            | `bool`     | Indicator whether the scraped timestamps should be respected. | `true` | no
+`track_timestamps_staleness`  | `bool`     | Indicator whether to track the staleness of the scraped timestamps. | `false` | no
```

Comment:
@bboreham Regarding #5921 - for now I intend to default this to `false` for a few reasons:
* IIUC, `track_timestamps_staleness = true` only makes sense if out of order ingestion is allowed in the back end database.
* Apparently there are implications to querying and [alerting](https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/).

Please let me know if you disagree. I'm open to changing this to default to `true` prior to merging. If it causes a backwards-incompatible change for the users, we'll have to mention it explicitly in our changelog and write docs with steps on what users need to change to get the behaviour they need (e.g. how to change their alerts or dashboards). 

We usually list backwards incompatible changes here: 
https://grafana.com/docs/agent/latest/flow/release-notes/

Reply:
> > IIUC, track_timestamps_staleness = true only makes sense if out of order ingestion is allowed in the back end database.
> 
> Not as far as I know. This is to fix the long-standing irritation where cAdvisor metrics linger on for 5 minutes after the pod has gone.
> 
> The blog you cite relates to the last change in staleness handling 5 years ago; I don't think it is relevant here.

@bboreham I think the assumption behind setting `track_timestamps_staleness = true` is that if the scraper didn't get any samples then that must be because there aren't any. But is this a good assumption to make in the general case? If a sample is exposed via explicit timestamps, and if there is no new value to report, then what sample should be exposed for its series the next time a scrape happens? Is the convention to just report the same value with a new timestamp?

I think it makes sense to default `track_timestamps_staleness` to `true` if:
* The convention is that the absence of a sample is considered enough evidence to decide that the series is stale.
* There is no need to enable out of order ingestion. This is so that we prevent a situation where a timestamp arrives late (e.g. because it takes a long time to generate) but it can't be ingested in the TSDB because there is already a staleness marker with a more recent timestamp.

This explicit timestamp feature seems like a way to "push" metrics.... so I'm not sure what assumptions are ok to make. I suspect that "push" systems like OTel don't just declare a series as stale if no samples were pushed in a certain time.

Reply:
Hi @bboreham, would you mind getting back to us on the comment above please?

Reply:
If no sample is supplied for a timestamp, PromQL (at query time) will use the preceding value up to 5 minutes old.
This creates a long-standing issue with cAdvisor (i.e. Kubelet container metrics).

It's got nothing to do with "push".

There is no expectation that explicit timestamps come out of order.  This never worked historically in Prometheus, and there is no reason to suppose people started sending them.

> But is this a good assumption to make in the general case?

Yes, it is the standard behaviour when exporters do not supply the timestamp.  Which is the vast majority.

Reply:
By the way, I don't think we should change this default in the middle of an 800-line PR doing other things.
I can make a separate PR to fix it.

Reply:
Ok, thank you, I'll take the `track_timestamps_staleness` parameter out of this PR. We can introduce it in a different PR. I don't want to add it now and change its default value later, because `prometheus.scrape` is a [stable](https://grafana.com/docs/agent/latest/stability/#stable) component and its defaults aren't meant to change often.